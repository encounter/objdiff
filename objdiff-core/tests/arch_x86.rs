use objdiff_core::{diff, diff::display::SymbolFilter, obj};

mod common;

#[test]
#[cfg(feature = "x86")]
fn read_x86() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/staticdebug.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?PrintThing@@YAXXZ").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_combine_sections() {
    let diff_config = diff::DiffObjConfig {
        combine_data_sections: true,
        combine_text_sections: true,
        ..Default::default()
    };
    let obj = obj::read::parse(include_object!("data/x86/rtest.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj.sections);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_64() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86_64/vs2022.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "?Dot@Vector@@QEAAMPEAU1@@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn display_section_ordering() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/basenode.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    let obj_diff =
        diff::diff_objs(Some(&obj), None, None, &diff_config, &diff::MappingConfig::default())
            .unwrap()
            .left
            .unwrap();
    let section_display =
        diff::display::display_sections(&obj, &obj_diff, SymbolFilter::None, false, false, false);
    insta::assert_debug_snapshot!(section_display);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_jumptable() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/jumptable.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?test@@YAHH@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

// Inferred size of functions should ignore symbols with specific prefixes
#[test]
#[cfg(feature = "x86")]
fn read_x86_local_labels() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/local_labels.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
}
