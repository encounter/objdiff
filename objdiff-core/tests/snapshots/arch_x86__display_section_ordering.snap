---
source: objdiff-core/tests/arch_x86.rs
expression: section_display
---
[
    SectionDisplay {
        id: ".text-0",
        name: ".text",
        size: 47,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 28,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-1",
        name: ".text",
        size: 65,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 29,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-2",
        name: ".text",
        size: 5,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 30,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-3",
        name: ".text",
        size: 141,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 31,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-4",
        name: ".text",
        size: 120,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 34,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-5",
        name: ".text",
        size: 378,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 37,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-6",
        name: ".text",
        size: 130,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 38,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-7",
        name: ".text",
        size: 123,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 39,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-8",
        name: ".text",
        size: 70,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 40,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-9",
        name: ".text",
        size: 90,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 41,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-10",
        name: ".text",
        size: 82,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 42,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-11",
        name: ".text",
        size: 336,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 43,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-12",
        name: ".text",
        size: 193,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 50,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-13",
        name: ".text",
        size: 544,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 53,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-14",
        name: ".text",
        size: 250,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 56,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-15",
        name: ".text",
        size: 89,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 57,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
    SectionDisplay {
        id: ".text-16",
        name: ".text",
        size: 119,
        match_percent: None,
        symbols: [
            SectionDisplaySymbol {
                symbol: 60,
                is_mapping_symbol: false,
            },
        ],
        kind: Code,
    },
]
